js/src/common/utils/ItemList.ts

```diff
@@ -1,35 +1,51 @@
-class Item {
-  content: any;
+class Item<T> {
+  content:
+    | T
+    | (T & {
+        /**
+         * Set when calling `.toArray()`
+         */
+        itemName: string;
+      });
   priority: number;
+
+  /**
+   * Set when calling `.toArray()`
+   */
   key?: number;
 
-  constructor(content: any, priority?: number) {
+  constructor(content: T, priority: number) {
     this.content = content;
     this.priority = priority;
   }
 }
 
+/**
+ * Options for new `replace` syntax.
+ *
+ * Removes the need for passing `null` to parameters.
+ */
+interface IItemListReplaceOptions<T> {
```

Comment:
I think we should export this, right? Same with the class above? Is there any reason not to?

Reply:
Yeah, there's no reason not to.

I don't know why we didn't used to export the Item class. Was there any reason?

Reply:
I have no idea 🤷

Reply:
> Yeah, there's no reason not to.
> 
> I don't know why we didn't used to export the Item class. Was there any reason?

The Item class is an implementation detail, and should not be instantiated outside of this file. Likewise, IMO, interfaces describing the structure of function inputs shouldn't be exported either.

Reply:
I agree with you on not exporting the Item, but I'm on the fence about exporting the interface. The interface purely serves a purpose in typing, so exporting it could allow areas of core and extensions to encapsulate the function and modify it as needed, while still ensuring their new function is correctly typed.

Without exporting it, some hacky-ness is needed around the use of `Parameters<ItemList["replace"]>[1]` or something similar.

Reply:
Good point, let's export the interface but not the item class then.

Reply:
This should be obsolete as we want to deprecate changing priority with replace.